etl_manager/extract_metadata.py

```diff
@@ -0,0 +1,409 @@
+        If True, adds 5 datetime columns: mojap_extraction_datetime,
+        mojap_start_datetime, mojap_end_datetime, mojap_latest_record, mojap_image_tag
+
+    include_objects (boolean):
+        If True, will include metadata for DB_TYPE_OBJECT columns as array<character>.
+        Leave False to ignore these columns - AWS DMS doesn't extract them.
+
+    Returns
+    -------
+    List of dicts
+        Contains data for all the columns in the table, ready to write to json
+    """
+    # This lookup is specific to Oracle data types
+    # Likely to need separate dictionaries for other data sources
+    type_lookup = {
+        "DB_TYPE_DATE": "datetime",
+        "DB_TYPE_TIMESTAMP": "datetime",
+        "DB_TYPE_TIMESTAMP_TZ": "datetime",
+        "DB_TYPE_CHAR": "character",
+        "DB_TYPE_CLOB": "character",
+        "DB_TYPE_BLOB": "character",
+        "DB_TYPE_VARCHAR": "character",
+        "DB_TYPE_LONG": "character",
+        "DB_TYPE_RAW": "character",
+        "DB_TYPE_OBJECT": "array<character>",
+    }
+    columns = []
+
+    if include_op_column:
+        columns.append(
+            {
+                "name": "op",
+                "type": "character",
+                "description": "Type of change, for rows added by ongoing replication.",
+                "nullable": True,
+                "enum": ["I", "U", "D"],
+            }
+        )
+
+    # Data types to skip. This list is specific to working with
+    # Amazon Database Migration Service and an Oracle database.
+    # These are the Oracle datatypes that DMS can't copy
+    if include_objects:
+        skip = [
+            "DB_TYPE_ROWID",
+            "DB_TYPE_BFILE",
+            "REF",
+            "ANYDATA",
+        ]
+    else:
+        skip = [
+            "DB_TYPE_ROWID",
+            "DB_TYPE_BFILE",
+            "REF",
+            "ANYDATA",
+            "DB_TYPE_OBJECT",
+        ]
+    # Main column info - cursor.description has 7 set columns:
+    # name, type, display_size, internal_size, precision, scale, null_ok
+    # Usually type is just a lookup, but decimal types need scale and precision too
+    for col in cursor.description:
+        if col[1].name not in skip:
+            columns.append(
+                {
+                    "name": col[0].lower(),
+                    "type": f"decimal({col[4]},{col[5]})"
+                    if col[1].name == "DB_TYPE_NUMBER"
+                    else type_lookup[col[1].name],
+                    "description": "",
+                    "nullable": bool(col[6]),
+                }
+            )
+
+    document_columns = [
+        {
+            "name": "mojap_document_path",
+            "type": "character",
+            "description": "The path to the document",
+            "nullable": True,
+        }
+    ]
+    derived_columns = [
+        {
+            "name": "mojap_extraction_datetime",
+            "type": "datetime",
+            "description": "",
```

Comment:
Should probably add descriptions here

Reply:
Added (now I've found out what they're all for!)